
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../examples/">
      
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.3">
    
    
      
        <title>API documentation - GENIE Python Module Documentation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.50c56a3b.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../css/custom.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#icaparser.icaparser" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="GENIE Python Module Documentation" class="md-header__button md-logo" aria-label="GENIE Python Module Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            GENIE Python Module Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              API documentation
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="GENIE Python Module Documentation" class="md-nav__button md-logo" aria-label="GENIE Python Module Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    GENIE Python Module Documentation
  </label>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Documentation Home Page
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../examples/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Examples
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    API documentation
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    API documentation
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser" class="md-nav__link">
    <span class="md-ellipsis">
      icaparser
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.add_gene_types" class="md-nav__link">
    <span class="md-ellipsis">
      add_gene_types()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.apply_mutation_classification_rules" class="md-nav__link">
    <span class="md-ellipsis">
      apply_mutation_classification_rules()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.cleanup_cosmic" class="md-nav__link">
    <span class="md-ellipsis">
      cleanup_cosmic()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.common_variant_filter" class="md-nav__link">
    <span class="md-ellipsis">
      common_variant_filter()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.explode_consequence" class="md-nav__link">
    <span class="md-ellipsis">
      explode_consequence()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.filter_positions_by_transcripts" class="md-nav__link">
    <span class="md-ellipsis">
      filter_positions_by_transcripts()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.filter_positions_by_variants" class="md-nav__link">
    <span class="md-ellipsis">
      filter_positions_by_variants()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.filter_variants_by_transcripts" class="md-nav__link">
    <span class="md-ellipsis">
      filter_variants_by_transcripts()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_aggregated_mutation_table" class="md-nav__link">
    <span class="md-ellipsis">
      get_aggregated_mutation_table()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_biotype_priority" class="md-nav__link">
    <span class="md-ellipsis">
      get_biotype_priority()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_clinvar" class="md-nav__link">
    <span class="md-ellipsis">
      get_clinvar()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_clinvar_max_significance" class="md-nav__link">
    <span class="md-ellipsis">
      get_clinvar_max_significance()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_consequences" class="md-nav__link">
    <span class="md-ellipsis">
      get_consequences()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_cosmic_max_sample_count" class="md-nav__link">
    <span class="md-ellipsis">
      get_cosmic_max_sample_count()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_data_sources" class="md-nav__link">
    <span class="md-ellipsis">
      get_data_sources()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_default_gene_type_map" class="md-nav__link">
    <span class="md-ellipsis">
      get_default_gene_type_map()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_default_mutation_aggregation_rules" class="md-nav__link">
    <span class="md-ellipsis">
      get_default_mutation_aggregation_rules()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_default_mutation_classification_rules" class="md-nav__link">
    <span class="md-ellipsis">
      get_default_mutation_classification_rules()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_dna_json_files" class="md-nav__link">
    <span class="md-ellipsis">
      get_dna_json_files()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_gene_type" class="md-nav__link">
    <span class="md-ellipsis">
      get_gene_type()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_genes" class="md-nav__link">
    <span class="md-ellipsis">
      get_genes()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_gnomad_exome_max_af" class="md-nav__link">
    <span class="md-ellipsis">
      get_gnomad_exome_max_af()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_gnomad_max_af" class="md-nav__link">
    <span class="md-ellipsis">
      get_gnomad_max_af()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_header" class="md-nav__link">
    <span class="md-ellipsis">
      get_header()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_header_scalars" class="md-nav__link">
    <span class="md-ellipsis">
      get_header_scalars()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_max_af" class="md-nav__link">
    <span class="md-ellipsis">
      get_max_af()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_multi_sample_positions" class="md-nav__link">
    <span class="md-ellipsis">
      get_multi_sample_positions()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_mutation_table_for_files" class="md-nav__link">
    <span class="md-ellipsis">
      get_mutation_table_for_files()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_mutation_table_for_position" class="md-nav__link">
    <span class="md-ellipsis">
      get_mutation_table_for_position()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_mutation_table_for_positions" class="md-nav__link">
    <span class="md-ellipsis">
      get_mutation_table_for_positions()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_onekg_max_af" class="md-nav__link">
    <span class="md-ellipsis">
      get_onekg_max_af()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_pipeline_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      get_pipeline_metadata()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_position_by_coordinates" class="md-nav__link">
    <span class="md-ellipsis">
      get_position_by_coordinates()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_positions" class="md-nav__link">
    <span class="md-ellipsis">
      get_positions()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_sample" class="md-nav__link">
    <span class="md-ellipsis">
      get_sample()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_strongest_vep_consequence_name" class="md-nav__link">
    <span class="md-ellipsis">
      get_strongest_vep_consequence_name()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_strongest_vep_consequence_priority" class="md-nav__link">
    <span class="md-ellipsis">
      get_strongest_vep_consequence_priority()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_strongest_vep_consequence_rank" class="md-nav__link">
    <span class="md-ellipsis">
      get_strongest_vep_consequence_rank()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_vep_consequence_for_rank" class="md-nav__link">
    <span class="md-ellipsis">
      get_vep_consequence_for_rank()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_vep_priority_for_consequence" class="md-nav__link">
    <span class="md-ellipsis">
      get_vep_priority_for_consequence()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_vep_rank_for_consequence" class="md-nav__link">
    <span class="md-ellipsis">
      get_vep_rank_for_consequence()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.split_multi_sample_json_file" class="md-nav__link">
    <span class="md-ellipsis">
      split_multi_sample_json_file()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.strip_json_file" class="md-nav__link">
    <span class="md-ellipsis">
      strip_json_file()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.strip_json_files" class="md-nav__link">
    <span class="md-ellipsis">
      strip_json_files()
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser" class="md-nav__link">
    <span class="md-ellipsis">
      icaparser
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.add_gene_types" class="md-nav__link">
    <span class="md-ellipsis">
      add_gene_types()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.apply_mutation_classification_rules" class="md-nav__link">
    <span class="md-ellipsis">
      apply_mutation_classification_rules()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.cleanup_cosmic" class="md-nav__link">
    <span class="md-ellipsis">
      cleanup_cosmic()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.common_variant_filter" class="md-nav__link">
    <span class="md-ellipsis">
      common_variant_filter()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.explode_consequence" class="md-nav__link">
    <span class="md-ellipsis">
      explode_consequence()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.filter_positions_by_transcripts" class="md-nav__link">
    <span class="md-ellipsis">
      filter_positions_by_transcripts()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.filter_positions_by_variants" class="md-nav__link">
    <span class="md-ellipsis">
      filter_positions_by_variants()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.filter_variants_by_transcripts" class="md-nav__link">
    <span class="md-ellipsis">
      filter_variants_by_transcripts()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_aggregated_mutation_table" class="md-nav__link">
    <span class="md-ellipsis">
      get_aggregated_mutation_table()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_biotype_priority" class="md-nav__link">
    <span class="md-ellipsis">
      get_biotype_priority()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_clinvar" class="md-nav__link">
    <span class="md-ellipsis">
      get_clinvar()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_clinvar_max_significance" class="md-nav__link">
    <span class="md-ellipsis">
      get_clinvar_max_significance()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_consequences" class="md-nav__link">
    <span class="md-ellipsis">
      get_consequences()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_cosmic_max_sample_count" class="md-nav__link">
    <span class="md-ellipsis">
      get_cosmic_max_sample_count()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_data_sources" class="md-nav__link">
    <span class="md-ellipsis">
      get_data_sources()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_default_gene_type_map" class="md-nav__link">
    <span class="md-ellipsis">
      get_default_gene_type_map()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_default_mutation_aggregation_rules" class="md-nav__link">
    <span class="md-ellipsis">
      get_default_mutation_aggregation_rules()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_default_mutation_classification_rules" class="md-nav__link">
    <span class="md-ellipsis">
      get_default_mutation_classification_rules()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_dna_json_files" class="md-nav__link">
    <span class="md-ellipsis">
      get_dna_json_files()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_gene_type" class="md-nav__link">
    <span class="md-ellipsis">
      get_gene_type()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_genes" class="md-nav__link">
    <span class="md-ellipsis">
      get_genes()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_gnomad_exome_max_af" class="md-nav__link">
    <span class="md-ellipsis">
      get_gnomad_exome_max_af()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_gnomad_max_af" class="md-nav__link">
    <span class="md-ellipsis">
      get_gnomad_max_af()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_header" class="md-nav__link">
    <span class="md-ellipsis">
      get_header()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_header_scalars" class="md-nav__link">
    <span class="md-ellipsis">
      get_header_scalars()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_max_af" class="md-nav__link">
    <span class="md-ellipsis">
      get_max_af()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_multi_sample_positions" class="md-nav__link">
    <span class="md-ellipsis">
      get_multi_sample_positions()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_mutation_table_for_files" class="md-nav__link">
    <span class="md-ellipsis">
      get_mutation_table_for_files()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_mutation_table_for_position" class="md-nav__link">
    <span class="md-ellipsis">
      get_mutation_table_for_position()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_mutation_table_for_positions" class="md-nav__link">
    <span class="md-ellipsis">
      get_mutation_table_for_positions()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_onekg_max_af" class="md-nav__link">
    <span class="md-ellipsis">
      get_onekg_max_af()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_pipeline_metadata" class="md-nav__link">
    <span class="md-ellipsis">
      get_pipeline_metadata()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_position_by_coordinates" class="md-nav__link">
    <span class="md-ellipsis">
      get_position_by_coordinates()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_positions" class="md-nav__link">
    <span class="md-ellipsis">
      get_positions()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_sample" class="md-nav__link">
    <span class="md-ellipsis">
      get_sample()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_strongest_vep_consequence_name" class="md-nav__link">
    <span class="md-ellipsis">
      get_strongest_vep_consequence_name()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_strongest_vep_consequence_priority" class="md-nav__link">
    <span class="md-ellipsis">
      get_strongest_vep_consequence_priority()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_strongest_vep_consequence_rank" class="md-nav__link">
    <span class="md-ellipsis">
      get_strongest_vep_consequence_rank()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_vep_consequence_for_rank" class="md-nav__link">
    <span class="md-ellipsis">
      get_vep_consequence_for_rank()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_vep_priority_for_consequence" class="md-nav__link">
    <span class="md-ellipsis">
      get_vep_priority_for_consequence()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.get_vep_rank_for_consequence" class="md-nav__link">
    <span class="md-ellipsis">
      get_vep_rank_for_consequence()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.split_multi_sample_json_file" class="md-nav__link">
    <span class="md-ellipsis">
      split_multi_sample_json_file()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.strip_json_file" class="md-nav__link">
    <span class="md-ellipsis">
      strip_json_file()
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#icaparser.icaparser.strip_json_files" class="md-nav__link">
    <span class="md-ellipsis">
      strip_json_files()
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  <h1>API documentation</h1>

<div class="doc doc-object doc-module">



<a id="icaparser.icaparser"></a>
  <div class="doc doc-contents first">
  
      <p>Parser for JSON files from Illumina Connected Annotations pipeline.</p>

  

  <div class="doc doc-children">










<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.add_gene_types" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">add_gene_types</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Adds the gene type to each transcript.</p>
<p>Transcripts will be annotated with the gene type (<em>oncogene</em>, <em>tsg</em>,
<em>mixed</em>) by adding a new attribute <code>geneType</code>. Only transcripts with
one of these three gene types get this additional annotation. Other
transcripts will not get the <code>geneType</code> attribute.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>positions</code></b>
        –
        <div class="doc-md-description">
          <p>list of filtered or unfiltered positions from JSON files</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A list of positions with additional annotation of transcripts</p>
      </div>
      </li>
  </ul>



<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">icaparser</span> <span class="k">as</span> <span class="nn">icap</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">positions</span> <span class="o">=</span> <span class="n">icap</span><span class="o">.</span><span class="n">add_gene_types</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span>
</span></code></pre></div>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.apply_mutation_classification_rules" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">apply_mutation_classification_rules</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">rule_set</span><span class="o">=</span><span class="n">get_default_mutation_classification_rules</span><span class="p">(),</span> <span class="n">gene_type_map</span><span class="o">=</span><span class="n">get_default_gene_type_map</span><span class="p">(),</span> <span class="n">hide_progress</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Applies mutation classification rules to all positions.</p>
<p>Each mutation is categorized for each isoform that overlaps with the genomic
position of the mutation. Each transcript which passed the "mutated" or
"uncertain" rule of the classification rules gets a new attribute
<code>mutation_status</code> with the value "mutated" or "uncertain".</p>
<p>In addition to classifying mutations based on their impact on transcript
isoforms, this function also assembles the mutation status on sample and
gene level without further aggregation. The impact depends on the type of
gene ("gof" or "lof"), so the impacts are assembled separately for each gene
type:</p>
<p><code>sample_id</code> → <code>gene</code> → <code>gene_type</code> → <code>variant_id</code> → <code>mutationStatus</code></p>
<p>The impact of a particular mutational variant can be different for different
overlapping transcript variants of a gene, and the transcript variants can
also belong to different genes. The strongest impact on any overlapping
transcript of a gene is defined as the impact of that mutational
variant on the gene. The analyst must decide which isoforms are used to
classify genes. For example, only canonical transcripts may be considered.
Alternatively, all transcripts or a subset of transcripts may be used.
Therefore, it is necessary to first apply transcript-level filters to all
genomic positions  before determining the mutation status of genes.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>positions</code></b>
        –
        <div class="doc-md-description">
          <p>list of positions</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>rule_set</code></b>
        –
        <div class="doc-md-description">
          <p>rules for classifying "gof" and "lof" genes. See also
<code>get_default_mutation_classification_rules()</code> for an example
if the default rule set needs to be modified.</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>gene_type_map</code></b>
        –
        <div class="doc-md-description">
          <p>dictionary for mapping gene types to canonical gene
types. See also <code>get_default_gene_type_map()</code> for an example
if the default mapping table needs to be modified.</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A list of positions and a dictionary with assembled and aggregated</p>
      </div>
      </li>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>mutations</p>
      </div>
      </li>
  </ul>



<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">icaparser</span> <span class="k">as</span> <span class="nn">icap</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">positions</span><span class="p">,</span> <span class="n">sample_muts</span> <span class="o">=</span> <span class="n">icap</span><span class="o">.</span><span class="n">apply_mutation_classification_rules</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span>
</span></code></pre></div>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.cleanup_cosmic" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">cleanup_cosmic</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Remove Cosmic entries with alleles not matching the variant alleles.</p>
<p>ICA attaches Cosmic entries to variants based on position only, which
leads to wrong assignments of Cosmic entries to variants. This function
removes all Cosmic entries from a variant for which reference and altered
alleles do not match those of the variant.</p>
<p>Filtering is done in place.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>positions</code></b>
        –
        <div class="doc-md-description">
          <p>list of positions to clean up</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A list of positions with cleaned up Cosmic entries</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.common_variant_filter" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">common_variant_filter</span><span class="p">(</span><span class="n">variant</span><span class="p">,</span> <span class="n">max_af</span><span class="o">=</span><span class="mf">0.01</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get a variant filter based on GnomAD, GnomAd Exome, and 1000 Genomes.</p>
<p>Returns True if none of the maximum allele frequencies from GnomAD,
GnomAD Exome and 1000 Genomes is larger than <code>max_af</code>.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>variant</code></b>
        –
        <div class="doc-md-description">
          <p>the variant to investigate</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>max_af</code></b>
        –
        <div class="doc-md-description">
          <p>the mixum allele frequences threshold</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A bool</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.explode_consequence" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">explode_consequence</span><span class="p">(</span><span class="n">mutation_table</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Explode the VEP consequence column of a mutation table.</p>
<p>Exploding the VEP consequence column with the standard Pandas
<code>explode()</code> function would return consquences as strings, not as
ordered categories. This function will instead return a consequence
columns which is an ordered category. The categories are ordered by
their impact.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>mutation_table</code></b>
        –
        <div class="doc-md-description">
          <p>the mutation table to explode</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>inplace</code></b>
        –
        <div class="doc-md-description">
          <p>if True, then modify the mutation_table in place instead
of returning a new object</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A pandas.DataFrame</p>
      </div>
      </li>
  </ul>



<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">icaparser</span> <span class="k">as</span> <span class="nn">icap</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">icap</span><span class="o">.</span><span class="n">explode_consequence</span><span class="p">(</span><span class="n">mutation_table</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</span></code></pre></div>
    <div class="highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">mutation_table_exploded</span> <span class="o">=</span> <span class="n">icap</span><span class="o">.</span><span class="n">explode_consequence</span><span class="p">(</span><span class="n">mutation_table</span><span class="p">)</span>
</span></code></pre></div>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.filter_positions_by_transcripts" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">filter_positions_by_transcripts</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">filter_func</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Filter positions based on a filter function for transcripts.</p>
<p>Apply a filter function to all transcripts of each position.
Transcripts not passing the filter are removed from a position.
Positions without any transcripts passing the filter are removed
from the returned list.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>positions</code></b>
        –
        <div class="doc-md-description">
          <p>list of positions to filter</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>filter_func</code></b>
        –
        <div class="doc-md-description">
          <p>function taking a transcript and returning a bool.
         True means to keep the transcript.</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A list</p>
      </div>
      </li>
  </ul>

<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>is_canonical_transcript = lambda x: x.get('isCanonical', False)
canonical_positions = icap.filter_positions_by_transcripts(
        non_common_positions,
        is_canonical_transcript
    )</p>
</blockquote>
</blockquote>
</blockquote>
</details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.filter_positions_by_variants" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">filter_positions_by_variants</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">filter_func</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Filter positions based on a filter function for variants.</p>
<p>Apply a filter function to all variants of each position.
Variants not passing the filter are removed from a position.
Positions without any variants passing the filter are removed
from the returned list.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>positions</code></b>
        –
        <div class="doc-md-description">
          <p>list of positions to filter</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>filter_func</code></b>
        –
        <div class="doc-md-description">
          <p>function taking a variant and returning a bool.
         True means to keep the variant.</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A list</p>
      </div>
      </li>
  </ul>

<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>import icaparser as icap
ax_af = 0.01
is_not_common_variant = lambda x: icap.common_variant_filter(x, max_af)
non_common_positions = icap.filter_positions_by_variants(
        positions,
        is_not_common_variant
    )</p>
</blockquote>
</blockquote>
</blockquote>
</details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.filter_variants_by_transcripts" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">filter_variants_by_transcripts</span><span class="p">(</span><span class="n">variants</span><span class="p">,</span> <span class="n">filter_func</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Filter variants based on a filter function for transcripts.</p>
<p>Apply a filter function to all transcripts of each variant.
Transcripts not passing the filter are removed from a variant.
Variants without any transcripts passing the filter are removed
from the returned list.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>variants</code></b>
        –
        <div class="doc-md-description">
          <p>list of variants to filter</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>filter_func</code></b>
        –
        <div class="doc-md-description">
          <p>function taking a transcript and returning a bool.
         True means to keep the transcript.</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A list</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_aggregated_mutation_table" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_aggregated_mutation_table</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">sample_muts</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mutation_classification_rules</span><span class="o">=</span><span class="n">get_default_mutation_classification_rules</span><span class="p">(),</span> <span class="n">mutation_aggregation_rules</span><span class="o">=</span><span class="n">get_default_mutation_aggregation_rules</span><span class="p">(),</span> <span class="n">gene_type_map</span><span class="o">=</span><span class="n">get_default_gene_type_map</span><span class="p">(),</span> <span class="n">hide_progress</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Returns a sample-gene-mutationStatus table.</p>
<p>This function applies mutation classification rules to all mutational
variants and aggregates the mutations according to the aggregation rules.
This results in a table with one row for each sample-gene pair. The table
contains several columns with impacts for "lof" and "gof" on allele level
and gene level and with one additional column with the maximum impact for
both allele and gene level.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>positions</code></b>
        –
        <div class="doc-md-description">
          <p>list of positions. If sample_muts is also specified, it is
assumed that the positions have already been processed previously by
<code>apply_mutation_classification_rules</code> and we do not have to run
mutation classification again.</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>sample_muts</code></b>
        –
        <div class="doc-md-description">
          <p>if <code>apply_mutation_classification_rules</code> has been run
before, you can use the second return value of that function as
the sample_muts argument. This is helpful for very large datasets
because otherwise <code>apply_mutation_classification_rules</code> will be
run again as an internal call within <code>get_aggregated_mutation_table</code>,
which is time consuming for very large data sets. This also means
that if <code>sample_muts</code> are specified, the <code>mutation_classification_rules</code>
argument is ignored and has no effect.</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>mutation_classification_rules</code></b>
        –
        <div class="doc-md-description">
          <p>rules for classifying single mutations.
See <code>get_default_mutation_classification_rules()</code> for details.</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>mutation_aggregation_rules</code></b>
        –
        <div class="doc-md-description">
          <p>rules for aggregation mutations.
See <code>get_default_mutation_aggregation_reles()</code> for details.</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>gene_type_map</code></b>
        –
        <div class="doc-md-description">
          <p>dictonary for mapping gene types to canonical gene types.
See <code>get_default_gene_type_map()</code> for details.</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A pandas.DataFrame with the mutation table.</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_biotype_priority" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_biotype_priority</span><span class="p">(</span><span class="n">biotype</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get the numeric priority of a biotype.</p>
<p>The numeric priority of a biotype that is returned by this function
is the same as defined by https://github.com/mskcc/vcf2maf/blob/master/vcf2maf.pl.
Biotypes are 'protein_coding', 'LRG_gene', ,'miRNA', ...</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>biotype</code></b>
        –
        <div class="doc-md-description">
          <p>a string with a single biotype</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>An integer, smaller values mean higher priority</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_clinvar" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_clinvar</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get a table of all ClinVar annotations for a variant.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>variant</code></b>
        –
        <div class="doc-md-description">
          <p>the variant to investigate</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A pandas.DataFrame</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_clinvar_max_significance" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_clinvar_max_significance</span><span class="p">(</span><span class="n">variant</span><span class="p">,</span> <span class="n">ordered_significances</span><span class="o">=</span><span class="n">_CLINVAR_ORDERED_SIGNIFICANCES</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get the maximum signifinance for all ClinVar annotations of a variant.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>variant</code></b>
        –
        <div class="doc-md-description">
          <p>the variant to investigate</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>ordered_significances</code></b>
        –
        <div class="doc-md-description">
          <p>ranked order of ClinVar significances</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A string</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_consequences" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_consequences</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get a list of consequences for a transcript</p>
<p>A list of consequences for a transcript is returned. If any of the annotated
consequences is a combination if single consequences, separated by
ampersands ('&amp;') or commas, such a consequence is split into single consequences.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>transcript</code></b>
        –
        <div class="doc-md-description">
          <p>the transcript to get the consequences for</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A list of strings</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_cosmic_max_sample_count" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_cosmic_max_sample_count</span><span class="p">(</span><span class="n">variant</span><span class="p">,</span> <span class="n">only_allele_specific</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get the maximum sample count for all Cosmic annotations of a variant.</p>
<p>A variant can have no, one or multiple associated Cosmic identifiers. This
function returns the maximum sample count of all Cosmic identifers. For each
Cosmic identifiers, sample numbers are summed up across all indications.
Returns 0 if no Cosmic identifier exists for this variant.</p>
<p>The 'only_allele_specific' argument is used to exclude Cosmic entries that
annotate the same chromosomal location but an allele that is different from
the allele of the annotated variant. ICA annotates a variant with all
Cosmic entries for that chromosomal location, irrespective of alleles. When
counting Cosmic samples, this leads to an overestimation of Cosmic sample
counts for a particular variant. Therefore, 'only_allele_specific' is True
by default to count only samples from Cosmic entries with matching alleles.
Occasionally, it may be desired, though, to count all samples with mutations
at a given position, irrespective of allele. For example, several different
alleles at a functional site of a gene can lead to function-disrupting
mutations, so we want to get the maximum sample count for any allele at that
position. One might also think of adding the sample counts for all Cosmic
entries annotating a variant, but this does not work  due to redundancy of
Cosmic entries. Older Cosmic versions often included the same sample in
different Cosmic entries. And newer Cosmic versions often have multiple
entries for an allele, one for each transcript variant, with the same
underlying samples.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>variant</code></b>
        –
        <div class="doc-md-description">
          <p>the variant to investigate</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>only_allele_specific</code></b>
        –
        <div class="doc-md-description">
          <p>consider only cosmic entries with alleles
                  matching the allele of the annotated variant</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>An integer</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_data_sources" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_data_sources</span><span class="p">(</span><span class="n">file</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Extract a table with annotation data sources from the JSON header.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file</code></b>
        –
        <div class="doc-md-description">
          <p>name of the ICA JSON file</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A pandas.DataFrame</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_default_gene_type_map" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_default_gene_type_map</span><span class="p">()</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Returns the default gene type map.</p>
<p>The canonical gene types are <code>gof</code>, <code>lof</code>, and the union of both.
Genes that need to be activated to drive a tumor are of type <code>gof</code>.
Genes that need to be deactivated to drive a tumor are of type <code>lof</code>.
Genes that need to be activated or deactivated depending on the context
are of the union of both types.
Genes for which it is unknown if they need to be activated or deactivated
are also annotated with both types.
Genes can be originally annotated with other type names than the canonical
ones. The gene type map is used to map these other gene type names to the
canonical gene types.</p>
<p>The default map is:</p>
<ul>
<li><code>oncogene</code> → <code>{"gof"}</code></li>
<li><code>tsg</code> → <code>{"lof"}</code></li>
<li><code>Act</code> → <code>{"gof"}</code></li>
<li><code>LoF</code> → <code>{"lof"}</code></li>
<li><code>mixed</code> → <code>{"gof", "lof"}</code></li>
<li><code>ambiguous</code> → <code>{"gof", "lof"}</code></li>
</ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A dictionary with mappings from gene types to canonical gene types</p>
      </div>
      </li>
  </ul>



<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">icaparser</span> <span class="k">as</span> <span class="nn">icap</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">icap</span><span class="o">.</span><span class="n">get_default_gene_type_map</span><span class="p">()</span>
</span></code></pre></div>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_default_mutation_aggregation_rules" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_default_mutation_aggregation_rules</span><span class="p">()</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Returns the default mutation aggregation rules.</p>
<p>Two types of the mutation status of a gene are introduced - allele level and
gene level:</p>
<ul>
<li>For "gof" genes (like oncogenes) it is sufficient if one of the alleles of
    one of the relevant isoforms has an activating mutation.</li>
<li>For "lof" genes (like tumor suppressor genes) all alleles of all relevant
    isoforms need to be functionally disrupted, either by mutations or by
    other means.</li>
<li>For Mixed and Other genes, the impact of a mutation is defined as the
    highest impact according to "gof" rules and "lof" rules.</li>
</ul>
<p>For "gof" genes, the allele and gene level classifications are identical
unless there is additional information about activating modifications other
than mutations. For "lof" genes, allele and gene level classifications
may be different. For example, a truncating mutation of a tumor suppressor
gene is functionally disruptive for the affected allele. However, there may
be other functionally active alleles of the same gene, i.e., the gene itself
may still be active. All alleles of such a gene must be dysfunctional,
either by additional mutations or by other processes, such as copy number
deletions or hypermethylation. Therefore, a single variant that is
disruptive at the allele level is not necessarily also disruptive at the
gene level. For "lof" genes, we usually do not have enough information
for a reliable estimation of functional effects. Instead, some heuristic
ules must be applied, and the analyst must decide whether to work with
allele-level or gene-level classifications. We designate a "lof" gene as
functionally disrupted (strong impact) if it has at least two mutations with
either strong impact or uncertain impact. If a "lof" gene has only one of
these mutations, it is designated as uncertain impact at the gene level,
even if one of these mutations has a strong impact at the allele level. By
categorizing effects at both allele and gene levels, we retain the
flexibility to decide in downstream analyses how to merge some of these
categories for subsequent statistical calculations.</p>
<p>Additional filters based on publications, white lists and black lists for
mutations can be applied. These are not part of the first version and will
be added later. White and black lists can be based, for example, on ClinVar
and on the publication by Hess et al. (2019).</p>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A dictionary</p>
      </div>
      </li>
  </ul>



<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">icaparser</span> <span class="k">as</span> <span class="nn">icap</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">icap</span><span class="o">.</span><span class="n">get_default_mutation_aggregation_rules</span><span class="p">()</span>
</span></code></pre></div>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_default_mutation_classification_rules" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_default_mutation_classification_rules</span><span class="p">(</span><span class="n">cosmic_threshold</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Returns the default rules for classifying mutations.</p>
<p>Defines the default rules for classifying mutations. The returned dictonary
has keys "gof" and "lof", and the respective values are the rule sets for
these gene types. A rule set is again a dictionary with the keys "mutated"
and "uncertain". The values for "mutated" or "uncertain" are dictionaries
with three filter functions, a "position_filter", a "variant_filter", and
a "transcript filter". For example, a transcript will be called "mutated"
if all three filters for "mutated" return True, and it will be called
"uncertain", if all three filter functions for "uncertain" return True.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>cosmic_threshold</code></b>
        –
        <div class="doc-md-description">
          <p>for "gof" genes, this is the "hotspot threshold" for
Cosmic, i.e., the minimum number of samples in Cosmic having that
mutation to consider a mutation a hot spot and, therefore, call the
mutation "mutated". If the number of Cosmic samples is smaller, the
mutation is called "uncertain".</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A dictionary with mutation classification rules</p>
      </div>
      </li>
  </ul>



<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">icaparser</span> <span class="k">as</span> <span class="nn">icap</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">icap</span><span class="o">.</span><span class="n">get_default_mutation_classification_rules</span><span class="p">()</span>
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">icap</span><span class="o">.</span><span class="n">get_default_mutation_classification_rules</span><span class="p">(</span><span class="n">cosmic_threshold</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
</span></code></pre></div>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_dna_json_files" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_dna_json_files</span><span class="p">(</span><span class="n">base_dir</span><span class="p">,</span> <span class="n">pattern</span><span class="o">=</span><span class="s1">&#39;*MergedVariants_Annotated_filtered.json.gz&#39;</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Find DNA annotation JSON files in or below <code>base_dir</code>.</p>
<p>Searches for ICA DNA annotation JSON files in and below <code>base_dir</code>.
All file names matching <code>pattern</code> are returned.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>base_dir</code></b>
        –
        <div class="doc-md-description">
          <p>base directory of directory subtree where to search
      for DNA annotation JSON files</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>pattern</code></b>
        –
        <div class="doc-md-description">
          <p>files names matching this pattern are returned</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A list of file names</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_gene_type" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_gene_type</span><span class="p">(</span><span class="n">gene_symbol</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get the gene type (oncogene, tsg, mixed) for a gene.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>gene_symbol</code></b>
        –
        <div class="doc-md-description">
          <p>the gene symbol of the gene</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>a string</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_genes" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_genes</span><span class="p">(</span><span class="n">file</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Extract gene annotation from a ICA JSON file.</p>
<p>The <code>genes</code> section of ICA JSON files is optional. If this section
is not included in the file, an empty list is returned.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file</code></b>
        –
        <div class="doc-md-description">
          <p>name of the ICA JSON file</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A list</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_gnomad_exome_max_af" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_gnomad_exome_max_af</span><span class="p">(</span><span class="n">variant</span><span class="p">,</span> <span class="n">cohorts</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;afr&#39;</span><span class="p">,</span> <span class="s1">&#39;amr&#39;</span><span class="p">,</span> <span class="s1">&#39;eas&#39;</span><span class="p">,</span> <span class="s1">&#39;nfe&#39;</span><span class="p">,</span> <span class="s1">&#39;sas&#39;</span><span class="p">])</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get the maximum allele frequency for gnomAD Exome.</p>
<p>Get the maximum allele frequences across all major cohorts annotated
by gnomAD, Exome  excluding bottleneck populations (Ashkenazy Jews and
Finish) and <em>other</em>.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>variant</code></b>
        –
        <div class="doc-md-description">
          <p>the variant to investigate</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>cohorts</code></b>
        –
        <div class="doc-md-description">
          <p>subpopulations to include</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A float</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_gnomad_max_af" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_gnomad_max_af</span><span class="p">(</span><span class="n">variant</span><span class="p">,</span> <span class="n">cohorts</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;afr&#39;</span><span class="p">,</span> <span class="s1">&#39;amr&#39;</span><span class="p">,</span> <span class="s1">&#39;eas&#39;</span><span class="p">,</span> <span class="s1">&#39;nfe&#39;</span><span class="p">,</span> <span class="s1">&#39;sas&#39;</span><span class="p">])</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get the maximum allele frequency for gnomAD.</p>
<p>Get the maximum allele frequences across all major cohorts annotated
by gnomAD, excluding bottleneck populations (Ashkenazy Jews and Finish)
and <em>other</em>.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>variant</code></b>
        –
        <div class="doc-md-description">
          <p>the variant to investigate</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>cohorts</code></b>
        –
        <div class="doc-md-description">
          <p>subpopulations to include</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A float</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_header" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_header</span><span class="p">(</span><span class="n">file</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Extract the header element from a ICA JSON file.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file</code></b>
        –
        <div class="doc-md-description">
          <p>name of the ICA JSON file</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A dictionary with the header from the JSON file</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_header_scalars" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_header_scalars</span><span class="p">(</span><span class="n">file</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Extract a table with all scalar attributes from the JSON header.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file</code></b>
        –
        <div class="doc-md-description">
          <p>name of the ICA JSON file</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A pandas.DataFrame</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_max_af" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_max_af</span><span class="p">(</span><span class="n">variant</span><span class="p">,</span> <span class="n">source</span><span class="p">,</span> <span class="n">cohorts</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get the maximum allele frequency for a particular annotation source.</p>
<p>Get the maximum allele frequency across all cohorts annotated by the
annotation source.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>variant</code></b>
        –
        <div class="doc-md-description">
          <p>the variant to investigate</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>source</code></b>
        –
        <div class="doc-md-description">
          <p>the annotation source to use</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>cohorts</code></b>
        –
        <div class="doc-md-description">
          <p>subpopulations to include; include all if cohorts == None</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A float</p>
      </div>
      </li>
  </ul>

<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>import icaparser as icap
icap.get_max_af(variant, 'gnomad')</p>
</blockquote>
</blockquote>
</blockquote>
</details>
  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_multi_sample_positions" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_multi_sample_positions</span><span class="p">(</span><span class="n">files</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Extract all positions for a set of ICA JSON files.</p>
<p>The sample id is stored as an additional new attribute of the
<code>samples</code> element of a position. The <code>samples</code> element is a list,
although ICA usually only creates single sample JSON files.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>files</code></b>
        –
        <div class="doc-md-description">
          <p>names of the ICA JSON files</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>args</code></b>
        –
        <div class="doc-md-description">
          <p>extra arguments forwarded to get_positions()</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>kwargs</code></b>
        –
        <div class="doc-md-description">
          <p>extra named arguments forwarded to get_positions()</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A list</p>
      </div>
      </li>
  </ul>



<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">icaparser</span> <span class="k">as</span> <span class="nn">icap</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">positions</span> <span class="o">=</span> <span class="n">icap</span><span class="o">.</span><span class="n">get_multi_sample_positions</span><span class="p">(</span><span class="n">json_files</span><span class="p">)</span>
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">positions</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;samples&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;sampleId&#39;</span><span class="p">])</span>
</span></code></pre></div>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_mutation_table_for_files" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_mutation_table_for_files</span><span class="p">(</span><span class="n">json_files</span><span class="p">,</span> <span class="n">max_af</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">min_vep_consequence_priority</span><span class="o">=</span><span class="mi">6</span><span class="p">,</span> <span class="n">min_cosmic_sample_count</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">only_canonical</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">extra_variant_filters</span><span class="o">=</span><span class="p">[],</span> <span class="n">extra_transcript_filters</span><span class="o">=</span><span class="p">[])</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get an annotated table of all filtered transcripts from a list of ICA JSON files.</p>
<p>Load all positions from a list of ICA JSON files and filter them.
Positions having any remaining variants and transcripts passing the filter
are returned as an annotated table.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>files</code></b>
        –
        <div class="doc-md-description">
          <p>list of ICA JSON files</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>max_af</code></b>
        –
        <div class="doc-md-description">
          <p>maximum allele frequency for gnomAD, gnomAD Exome and 1000 Genomes.
    Only variants with maximum allele frequencies below this threshold
    will be returned.</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>min_vep_consequence_priority</code></b>
        –
        <div class="doc-md-description">
          <p>only transcripts with a minimum VEP
    consequence priority not larger than this threshold will be retained</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>min_cosmic_sample_count</code></b>
        –
        <div class="doc-md-description">
          <p>only variants with a maximum cosmic sample count
    not lower than this threshold will be retained</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>only_canonical</code></b>
        –
        <div class="doc-md-description">
          <p>if true, only canonical transcripts will be retained</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>extra_variant_filters</code></b>
        –
        <div class="doc-md-description">
          <p>any additional filters to apply to variants.
    Filters shall return True to keep a variant.</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>extra_transcript_filters</code></b>
        –
        <div class="doc-md-description">
          <p>any additional filters to apply to transcripts.
    Filters shall return True to keep a transcript.</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A pandas.DataFrame</p>
      </div>
      </li>
  </ul>



<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">icaparser</span> <span class="k">as</span> <span class="nn">icap</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">extra_transcript_filters</span> <span class="o">=</span> <span class="p">[</span>
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="go">        lambda x: x.get(&#39;source&#39;, &#39;&#39;) == &#39;Ensembl&#39;,</span>
</span><span id="__span-0-4"><a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="go">        lambda x: x.get(&#39;hgnc&#39;, &#39;&#39;) == &#39;KRAS&#39;</span>
</span><span id="__span-0-5"><a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="go">    ]</span>
</span><span id="__span-0-6"><a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">mut_table</span> <span class="o">=</span> <span class="n">icap</span><span class="o">.</span><span class="n">get_mutation_table_for_files</span><span class="p">(</span>
</span><span id="__span-0-7"><a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="go">        json_files,</span>
</span><span id="__span-0-8"><a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="go">        extra_transcript_filters=extra_transcript_filters</span>
</span><span id="__span-0-9"><a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="go">    )</span>
</span></code></pre></div>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_mutation_table_for_position" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_mutation_table_for_position</span><span class="p">(</span><span class="n">position</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get an annotated table of all transcripts.</p>
<p>Returns an annotated table of all transcripts that are affected
by a mutation at a position.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>position</code></b>
        –
        <div class="doc-md-description">
          <p>the position to investigate</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A pandas.DataFrame</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_mutation_table_for_positions" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_mutation_table_for_positions</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">hide_progress</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get an annotated table of all transcripts for all positions.</p>
<p>Returns an annotated table of all transcripts that are affected
by a mutation at any of the position.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>positions</code></b>
        –
        <div class="doc-md-description">
          <p>the positions to investigate</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A pandas.DataFrame</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_onekg_max_af" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_onekg_max_af</span><span class="p">(</span><span class="n">variant</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get the maximum allele frequency for the 1000 Genomes Project.</p>
<p>Get the maximum allele frequences across all cohorts annotated
by the 1000 Genomes Project.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>variant</code></b>
        –
        <div class="doc-md-description">
          <p>the variant to investigate</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A float</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_pipeline_metadata" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_pipeline_metadata</span><span class="p">(</span><span class="n">files</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Extract a table with metadata annotation pipeline run from the JSON header.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file</code></b>
        –
        <div class="doc-md-description">
          <p>name of the ICA JSON file</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A pandas.DataFrame</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_position_by_coordinates" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_position_by_coordinates</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="n">chromosome</span><span class="p">,</span> <span class="n">position</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Extract a particular position from a position list.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>positions</code></b>
        –
        <div class="doc-md-description">
          <p>a list of positions</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>chromosome</code></b>
        –
        <div class="doc-md-description">
          <p>name of the chromosome</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>position</code></b>
        –
        <div class="doc-md-description">
          <p>numeric posion on the chromosome</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A dict</p>
      </div>
      </li>
  </ul>



<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">icaparser</span> <span class="k">as</span> <span class="nn">icap</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">icap</span><span class="o">.</span><span class="n">get_position_by_coordinates</span><span class="p">(</span><span class="n">positions</span><span class="p">,</span> <span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="mi">204399064</span><span class="p">)</span>
</span></code></pre></div>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_positions" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_positions</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">variant_filters</span><span class="o">=</span><span class="p">[],</span> <span class="n">transcript_filters</span><span class="o">=</span><span class="p">[])</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Extract all positions from a ICA JSON file.</p>
<p>The sample id is stored as an additional new attribute of the
<code>samples</code> element of a position. The <code>samples</code> element is a list,
although ICA usually only creates single sample JSON files.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file</code></b>
        –
        <div class="doc-md-description">
          <p>name of the ICA JSON file</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>variant_filters</code></b>
        –
        <div class="doc-md-description">
          <p>any filters to apply to variants.
    Filters shall return True to keep a variant.</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>transcript_filters</code></b>
        –
        <div class="doc-md-description">
          <p>any filters to apply to transcripts.
    Filters shall return True to keep a transcript.</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A list</p>
      </div>
      </li>
  </ul>



<p><strong>Examples:</strong></p>
    <div class="highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">transcript_filters</span> <span class="o">=</span> <span class="p">[</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="go">        lambda x: x.get(&#39;source&#39;, &#39;&#39;) == &#39;Ensembl&#39;,</span>
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="go">        lambda x: x.get(&#39;hgnc&#39;, &#39;&#39;) == &#39;KRAS&#39;</span>
</span><span id="__span-0-4"><a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="go">    ]</span>
</span><span id="__span-0-5"><a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="gp">&gt;&gt;&gt; </span><span class="n">positions</span> <span class="o">=</span> <span class="n">icap</span><span class="o">.</span><span class="n">get_sample_positions</span><span class="p">(</span>
</span><span id="__span-0-6"><a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="go">        json_file,</span>
</span><span id="__span-0-7"><a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="go">        transcript_filters = transcript_filters</span>
</span><span id="__span-0-8"><a id="__codelineno-0-8" name="__codelineno-0-8" href="#__codelineno-0-8"></a><span class="go">    )</span>
</span><span id="__span-0-9"><a id="__codelineno-0-9" name="__codelineno-0-9" href="#__codelineno-0-9"></a><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">positions</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;samples&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;sampleId&#39;</span><span class="p">])</span>
</span></code></pre></div>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_sample" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_sample</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="n">suffix</span><span class="o">=</span><span class="s1">&#39;(-D[^.]*)?</span><span class="se">\\</span><span class="s1">.bam&#39;</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Extract the sample name from a ICA JSON file.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>file</code></b>
        –
        <div class="doc-md-description">
          <p>name of the ICA JSON file</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>suffix</code></b>
        –
        <div class="doc-md-description">
          <p>regular expression to remove from the sample name in the JSON
file. Defaults to '(-D[^.]*)?.bam'.</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A string with the name of the sample annotated in the JSON file</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_strongest_vep_consequence_name" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_strongest_vep_consequence_name</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get the name of the strongest VEP consequence for a transcript.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>transcript</code></b>
        –
        <div class="doc-md-description">
          <p>the transcript to investigate</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A string</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_strongest_vep_consequence_priority" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_strongest_vep_consequence_priority</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get the strongest priority of VEP consequence for a transcript.</p>
<p>Get the strongest numeric priority of all VEP consequences
for a transcript. Smaller numeric priorities mean stronger impact.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>transcript</code></b>
        –
        <div class="doc-md-description">
          <p>the transcript to investigate</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>An integer</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_strongest_vep_consequence_rank" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_strongest_vep_consequence_rank</span><span class="p">(</span><span class="n">transcript</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get the strongest rank of VEP consequences for a transcript.</p>
<p>Get the strongest numeric rank of all VEP consequences
for a transcript. Smaller ranks mean stronger impact.</p>
<p>The priority of consequences is taken into account first. So if
two consequences have different priorities, the consequence with
the higher priority (lower priority number) will be used, and the
rank for this consequence will be returned. If there are multiple
consequences with the same priority, the lowest (strongest) rank
will be returned.</p>
<p>For clarification: ranks are unique, i.e. all VEP consequences
ordered as listed on the VEP documentation page get the row number
of this table assigned as rank.</p>
<p>However, several consequences can have the same priority (e.g., stop
gained and frameshift have the same priority). Priorities are copied
from vcf2maf.pl.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>transcript</code></b>
        –
        <div class="doc-md-description">
          <p>the transcript to investigate</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>An integer</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_vep_consequence_for_rank" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_vep_consequence_for_rank</span><span class="p">(</span><span class="n">rank</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get the VEP consequence term of a numeric rank.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>rank</code></b>
        –
        <div class="doc-md-description">
          <p>the numeric rank of the consequence term</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>A string</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_vep_priority_for_consequence" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_vep_priority_for_consequence</span><span class="p">(</span><span class="n">consequence</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get the numeric priority of a VEP consequence term.</p>
<p>The numeric priority of a consequence that is returned by this function
  is the same as defined by https://github.com/mskcc/vcf2maf/blob/master/vcf2maf.pl.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>consequence</code></b>
        –
        <div class="doc-md-description">
          <p>the consequence term of the mutation</p>
        </div>
      </li>
  </ul>
      <p>Returns:
      An integer</p>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.get_vep_rank_for_consequence" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">get_vep_rank_for_consequence</span><span class="p">(</span><span class="n">consequence</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Get the numeric rank of a VEP consequence term.</p>
<p>The numeric rank of a consequence that is the position of the consequence
in this list of consequences at https://www.ensembl.org/info/genome/variation/prediction/predicted_data.html</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>consequence</code></b>
        –
        <div class="doc-md-description">
          <p>the consequence term of the mutation</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>An integer</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.split_multi_sample_json_file" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">split_multi_sample_json_file</span><span class="p">(</span><span class="n">json_file</span><span class="p">,</span> <span class="n">output_dir</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Splits a multi-sample JSON file into sample specifc JSON files.</p>
<p>This function reads a multi-sample JSON file that was generated by
annotating a multi-sample VCF file with ICA and splits it into
sample-specific JSON files.</p>
<p>Annotating very many single-sample VCF files with ICA is very time
consuming, because ICA reads all annotation sources for each VCF file
and this is dominating the runtime of ICA. It is therefore helpful to
first merge many single-sample VCF files into one or a small number of
multi-sample VCF files (for example, with <code>bcftools merge</code>), to annotate
the multi-sample VCF file with ICA, and then to split the multi-sample
JSON output of ICA into single-sample JSON files. These single-sample
JSON files are required for the rest of this package.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>json_file</code></b>
        –
        <div class="doc-md-description">
          <p>the multi-sample json input file</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>output_dir</code></b>
        –
        <div class="doc-md-description">
          <p>the directory where to write the single sample JSON files.
The directory will be created if it does not exist.</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>Nothing</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.strip_json_file" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">strip_json_file</span><span class="p">(</span><span class="n">ifname</span><span class="p">,</span> <span class="n">ofname</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Reduce the JSON file size by keeping only 'PASS' variants.</p>
<p>JSON files from Illumina's ICA pipeline can be very large because they
contain any deviation from the reference genome, irrespective of the
quality of the mutation call. Gzip compressed JSON files with sizes in the
gigabyte range cannot be processed by JSON packages that read the entire
file into memory. It is necessary to first reduce the size of JSON files by
removing all variants that do not meet Illumina's quality criteria.</p>
<p>This function reads a single JSON file and creates a single JSON outpout
file by removing all variants that do not pass Illumina's quality criteria.</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>ifname</code></b>
        –
        <div class="doc-md-description">
          <p>name of the  input file</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>ofname</code></b>
        –
        <div class="doc-md-description">
          <p>name of the output file</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>Nothing</p>
      </div>
      </li>
  </ul>

  </div>

</div>


<div class="doc doc-object doc-function">




<h2 id="icaparser.icaparser.strip_json_files" class="doc doc-heading">
          <code class="highlight language-python"><span class="n">strip_json_files</span><span class="p">(</span><span class="n">source_dir</span><span class="p">,</span> <span class="n">target_dir</span><span class="p">,</span> <span class="n">pattern</span><span class="o">=</span><span class="s1">&#39;*.json.gz&#39;</span><span class="p">)</span></code>

</h2>


  <div class="doc doc-contents ">
  
      <p>Strip all JSON files of a project by keeping only 'PASS' variants.</p>
<p>JSON files from Illumina's ICA pipeline can be very large because they
contain any deviation from the reference genome, irrespective of the
quality of the mutation call. Gzip compressed JSON files with sizes in the
gigabyte range cannot be processed by JSON packages that read the entire
file into memory. It is necessary to first reduce the size of JSON files by
removing all variants that do not meet Illumina's quality criteria.</p>
<p>This function searches <code>source_dir</code> recursively for all files matching the
<code>file_pattern</code>. Each of those files is processed and a stripped version
keeping only variants that PASS Illumina's quality criteria is created. The
output file has the same name as the input file. The directory structure
below <code>source_dir</code> is replicated in <code>target_dir</code>. Output files get the
suffix '_filtered.json.gz'.‚</p>



  <p>Parameters:</p>
  <ul>
      <li class="field-body">
        <b><code>source_dir</code></b>
        –
        <div class="doc-md-description">
          <p>directory where to search for input JSON files</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>target_dir</code></b>
        –
        <div class="doc-md-description">
          <p>directory where to save the stripped outpout JSON files</p>
        </div>
      </li>
      <li class="field-body">
        <b><code>file_pattern</code></b>
        –
        <div class="doc-md-description">
          <p>files matching this pattern will be processed</p>
        </div>
      </li>
  </ul>



  <p>Returns:</p>
  <ul>
      <li class="field-body">
        –
        <div class="doc-md-description">
          <p>Nothing</p>
      </div>
      </li>
  </ul>

<details class="example" open>
  <summary>Example</summary>
  <blockquote>
<blockquote>
<blockquote>
<p>strip_json_files('../Data/Original', '../Data/Derived')</p>
</blockquote>
</blockquote>
</blockquote>
</details>
  </div>

</div>



  </div>

  </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["content.code.copy"], "search": "../assets/javascripts/workers/search.f886a092.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.d7c377c4.min.js"></script>
      
    
  </body>
</html>